Block Library: Add features to the Post Date block.

epiqueras opened the pull request:
Closes #19697

## Description

This PR implements the design from #19697.

## How to test this?

Verify all the features from #19697 work as expected with a Post Date block inserted in a regular post.

## Screenshots

<img width="738" alt="Screen Shot 2020-01-23 at 8 57 45 PM" src="https://user-images.githubusercontent.com/19157096/73039154-0660b500-3e23-11ea-90aa-1a7f505ca300.png">

## Types of Changes

*New Feature*: The Post Date block now has a more rich editing experience and allows you to customize the block-specific date format directly.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

epiqueras committed:
Block Library: Add features to the Post Date block.

shaunandrews commented:
The pencil icon leaves a lot to be desired. Its not at all obvious, and actually seems to imply "change" or "edit" — not "format". Maybe we can replace it with a label:

<img width="403" alt="image" src="https://user-images.githubusercontent.com/191598/74544687-8e634780-4f15-11ea-9b37-888ce8bc139a.png">

I wish there was more control over the date and time formatting. Combining them both into a single selection reduces the control — or, if we add more, requires more options to provide more control. If we were to separate the formatting we could offer more control without having to list a dozen different combinations:

<img width="198" alt="image" src="https://user-images.githubusercontent.com/191598/74544350-fbc2a880-4f14-11ea-93f7-dd8d2c58049e.png">

I found it jarring at first that selecting the block instantly shows the calendar to change the date. I wasn't expecting it, and I didn't really understand why that control was showing. Part of me thinks it should be something more intentional, like a "change" button in the toolbar:

<img width="400" alt="image" src="https://user-images.githubusercontent.com/191598/74544460-2a408380-4f15-11ea-86f3-e2c0731406f5.png">

epiqueras commented:
Using labels in the toolbar like that is not a good idea because
localization can break the layout.

shaunandrews commented:
Sure, I can see how translations could cause troubles here. But I also don't think a pencil icon is a good way to indicate "change the date and time formatting". I'm not sure such an icon exists.

We had a similar problem for the Image block, and ended up going with a "Replace" label after trying a handful of different icons.

epiqueras commented:
Agreed. I suggested we use a calendar icon with some sort of pencil/tool
overlaying on top.

shaunandrews commented:
> calendar icon with some sort of pencil/tool

The block icon itself is already a calendar, and I worry about the repetition. Also, would this icon be for changing the date/time formatting, or editing the date of the post?

Here's a few quick variations of the calendar, pencil, and clock icons:

<img width="344" alt="image" src="https://user-images.githubusercontent.com/191598/74548461-e9983880-4f1b-11ea-995b-60af59330b60.png">

I don't think any of them are better or more obvious that using a label for the formatting. Generally, combining multiple icons at such a small size is not recommended — both icons tend to lose their "shape" and become unrecognizable.

I could, however, see the pencil icon working for changing the date of the post.

<img width="270" alt="image" src="https://user-images.githubusercontent.com/191598/74548638-3c71f000-4f1c-11ea-891a-e01eca015936.png">

epiqueras commented:
How about the pencil for changing the date and a calendar or placeholder
date with a gear icon on top for the format.

mtias commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -3,24 +3,67 @@
+				</Popover>
+			) }
+			<BlockControls>
+				<ToolbarGroup
```
I don't think the format is changed often enough to warrant a placement in the toolbar. It could be just an inspector control.

epiqueras replied:
cc @shaunandrews

epiqueras committed:
Block Library: Redesign new Post Date block features.

epiqueras commented:
I've refactored the opening of the calendar popover so that it now opens through a pencil icon on the toolbar.

I also moved the format dropdown to the inspector. This solves the icon problem for now.

@shaunandrews I think this is ready to go.

ockham commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -2,25 +2,93 @@
+			.replace( /\\\\/g, '' )
+			.split( '' )
+			.reverse()
+			.join( '' )
```
Can we maybe add an example comment illustrating what `settings` looks like before and after transformation? The transformation is a bit abstract otherwise :sweat_smile:

epiqueras replied:
Yeah, done.

ockham commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -2,25 +2,93 @@
+		( formatKey ) => ( {
+			key: settings.formats[ formatKey ],
+			name: dateI18n( settings.formats[ formatKey ], date ),
+		} )
```
Might be a good fit for `_.map()` since we're only using `settings.formats`' values? (Or `Object.values().map()`, in case we're polyfilling for older browsers?)

epiqueras replied:
Sure, that makes sense.

epiqueras committed:
Comments and Style

ockham reviewed (commented):
Date/time selection seems to work, but I'm not sure how to change the format:

![post-date](https://user-images.githubusercontent.com/96308/75142315-018e6a00-571d-11ea-9611-65786703cbec.gif)

Edit: Ohhh, it's mentioned here: https://github.com/WordPress/gutenberg/pull/19857#issuecomment-587117676

> I also moved the format dropdown to the inspector.

Will test the format now.

ockham commented:
This works, but I find the split (data/time selection via block toolbar, format selection via block inspector) rather unintuitive, esp for a block that basically has only those two customization options :thinking: 

![post-date2](https://user-images.githubusercontent.com/96308/75142898-3cdd6880-571e-11ea-9b23-b1811136037f.gif)

I think from a user's POV, it begs the question, why aren't both options in the same place?

ockham reviewed (approved):
Code-wise, this is looking good.

I have some [UI concerns](https://github.com/WordPress/gutenberg/pull/19857#issuecomment-590243469), but it's probably okay to address those in a future iteration.

mtias commented:
We should avoid the pencil as the "edit" icon.

mcsf commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -2,25 +2,95 @@
-
+	// To know if the current time format is a 12 hour time, look for "a".
+	// Also make sure this "a" is not escaped by a "/".
+	const is12Hour = /a(?!\\)/i.test(
+		settings.formats.time
+			.toLowerCase() // Test only for the lower case "a".
+			.replace( /\\\\/g, '' ) // Replace "//" with empty strings.
+			.split( '' )
+			.reverse()
+			.join( '' ) // Reverse the string and test for "a" not followed by a slash.
```
* Why reverse the string and not use a non-capturing group like `/(?:[^\\])a/`?
* If the RegExp already has the `i` flag, why pipe through `.toLowerCase`?

mcsf replied:
I now see that the original occurrence came with 256badd16f7bd6a10d7f6f66780862b8860c18b2, three years ago, so ignore this message.
